### ISSUE TITLE
Scrolling problem in Chromium-based apps when touching FP reader

### ISSUE THREAD

**Issue by EetuRasilainen** [gardiol/device_xiaomi_surya#1]
I have noticed some weird scrolling behavior on LineageOS in my web browser. Depending on how you are holding the device, kinetic scrolling might not be working or stopping randomly. This is caused by (accidentally) touching the fingerprint reader during scrolling.

You can verify this with any key event tester app (such as KeyTester). Touching the FP reader emits a key event (id 96) and apparently Chromium-based apps do not like that. This might explain that some people have scrolling issues and some people do not - it really depends on how you are holding your device, which also depends on the size of your hand :smile: 

The same thing used to happen in early ArrowOS builds a while ago and they were able to fix this.

Just in case it helps, these are the related ArrowOS commits which fix this issue in their ROMs:

(For pre-OSS builds):
https://github.com/ArrowOS-Devices/android_device_xiaomi_surya/commit/b79c334e53c0a350f111cf0489dab170c174f878

(For OSS-builds):
https://github.com/ArrowOS-Devices/android_device_xiaomi_surya/commit/2c4898110186a0e66c6e4aff9749539daec2b263

Edit: It seems they have patched `fingerprint.fpc.default.so` for this. This is the version they are using:

https://github.com/ArrowOS-Devices/android_vendor_xiaomi_sm6250-common/blob/arrow-11.0/proprietary/vendor/lib64/hw/fingerprint.fpc.default.so

**Comment by gardiol** [comment]
I understand this is a problem only for FPC devices.
Unfortunately, i cannot pin an unknown blob nor i can accept it without knowing what changes have been performed in the source code. 

I will look into alternate ways to achieve the same result, but frankly i am inclined to close this as not a ROM bug but applications bug.

**Comment by EetuRasilainen** [comment]
What do you mean by FPC devices? I think this affects both Poco X3 NFC and Poco X3 devices (surya and karna).

From what I understand @nullbytepl made a small edit to the original library using a hex editor and added a NOOP for report_input_event.

The very first fix for this from the ArrowOS folks did not involve any custom FW blobs though, see this commit:

https://github.com/ArrowOS-Devices/android_device_xiaomi_surya/commit/b79c334e53c0a350f111cf0489dab170c174f878

However a change within android_frameworks_native was required and I am not sure if the LineageOS devs would like that :thinking:

**Comment by gardiol** [comment]
Surya&Karna have two different FP sensors: the FPC or the Goodix. All my suryas have Goodix sensor not FPC. This bug affects only FPC sensors not Goodix sensors, this is what i mean.

I agree that the original way would not suit LOS approach to life.

Still, i believe this is an issue with the specific apps. Does it happen with Firefox for example?

Can you try to drop-in the Arrow OS modified library on top of the one in LOS and see if that by itself solves it?

**Comment by EetuRasilainen** [comment]
> Surya&Karna have two different FP sensors: the FPC or the Goodix. All my suryas have Goodix sensor not FPC. This bug affects only FPC sensors not Goodix sensors, this is what i mean.

Very interesting, did not know that. Thx.

> Still, i believe this is an issue with the specific apps. Does it happen with Firefox for example?

It does not happen with Firefox. It happens in all WebKit-based browsers and Android's System WebView. The latter is a core component of the Android system which cannot be replaced by Firefox. So unfortunately this is not an app-specific issue. All apps which use Android's WebView are affected by this bug (millions of apps).

> Can you try to drop-in the Arrow OS modified library on top of the one in LOS and see if that by itself solves it?

Just did that. Unfortunately this did not work. I am using the patched library but I'm still getting scrolling issues when touching the FP. That is weird. This is how my patched /vendor looks after a reboot:

```bash
surya:/ # sha1sum /vendor/lib64/hw/fingerprint.fpc.default.so                                                                                                                                                                                                                           
645bf256da2db3ce60dddab9b3282ab7241bbcb0  /vendor/lib64/hw/fingerprint.fpc.default.so
```
This is the sha1 from https://github.com/ArrowOS-Devices/android_device_xiaomi_surya/commit/2c4898110186a0e66c6e4aff9749539daec2b263 so I think I did the right thing. Strange, I don't understand that :thinking: .

How do I know whether I am on an FPC or goodix sensor?
```bash
surya:/ # dmesg |grep fpc                                                                                                                                                                                                                                                               
[    0.668358] Macle fpc1020_probeing failed as fpsensor=2(1=fp)
[    0.668365] fpc1020: probe of soc:fpc1020 failed with error -12
[    0.668418] fpc1020_init OK
surya:/ # dmesg |grep good                                                                                                                                                                                                                                                              
[    0.668683] input: uinput-goodix as /devices/virtual/input/input2
[    6.633485] QSEECOM: qseecom_load_app: App (goodixfp) does'nt exist, loading apps for first time
[    6.682942] QSEECOM: qseecom_load_app: App with id 4 (goodixfp) now loaded
```

When touching the FP sensor I am getting events on `/dev/input/event2`.

**Comment by gardiol** [comment]
Yes you have FPC sensor! (fpc1020_init OK) 
This is indeed a bug in the sensor management in the vendor part, since the Googix behaves correctly. Not uncommon on XIAOMI stuff. 

The point is how to fix it. If the replaced library does not fix it, then it means that the framework counterpart too is affected (the other commits from ArrowOS). This is weird and quite annoying.

I will try to think something else.

**Comment by EetuRasilainen** [comment]
Okay, cool.

I guess the line `input: uinput-goodix as /devices/virtual/input/input2` should not worry me, or should it? I know this is the virtual sysfs device path which should not correlate to the device nodes under `/dev/input/event*`. So it must be a coincidence that `/dev/input/event2` is actually spitting out events on FP touches. What I am saying is: can we be sure that uinput-goodix does not interfere with FPC devices?

Does your goodix reader also spit out events when you do a stupid `cat /dev/input/event2` while touching the FP reader?

**Comment by gardiol** [comment]
Yes, /dev/input/event2 is the standard path for the input device related to the fingerprint, that is correct.
Oddly enough, the FPC and the Goodix share the same manufacturer (Goodix) hence the confusion. 
They still are two different pieces of hardware.

**Comment by EetuRasilainen** [comment]
This is fixed as of `lineage-18.1-20220110-nightly-surya-signed.zip`. Thank you so much for looking into that. :+1: :+1: 

I think it would be cool if your fix could also be merged in the Poco X3 Pro branch of LineageOS (vayu). I know you have nothing to do with that device, but it's the same FP hardware with the same bug (my dad has it). Maybe you have an easy way of pinging the device maintainer behind the scenes. But you may also just close this issue now as you have done a great job already :smile:

**Comment by gardiol** [comment]
Good. The fix is at kernel level.
